File: code/agents/IngestionAgent.java
package agents;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.Timer;
import java.util.TimerTask;

import shared.AgentConfig;
import shared.StorageHelper;

/**
 * IngestionAgent autonomously collects data from a configured source URL and saves it to storage.
 * Usage: Instantiate with AgentConfig and call start().
 */
public class IngestionAgent {

    private final AgentConfig config;

    public IngestionAgent(AgentConfig config) {
        this.config = config;
    }

    public void start() {
        Timer timer = new Timer();
        timer.scheduleAtFixedRate(new IngestTask(), 0, config.getIngestionIntervalMs());
    }

    class IngestTask extends TimerTask {
        @Override
        public void run() {
            try {
                System.out.println("IngestionAgent: Fetching data from source...");
                URL url = new URL(config.getSourceUrl());
                BufferedReader reader = new BufferedReader(new InputStreamReader(url.openStream()));

                StringBuilder rawData = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    rawData.append(line).append("\n");
                }
                reader.close();

                StorageHelper.saveRawData(config.getStorageOutputPath(), rawData.toString());
                System.out.println("IngestionAgent: Data ingested and saved.");
            } catch (Exception e) {
                System.err.println("IngestionAgent: Failed to ingest data: " + e.getMessage());
            }
        }
    }

    /**
     * Standalone entry point for running the IngestionAgent independently.
     */
    public static void main(String[] args) {
        AgentConfig config = new AgentConfig("../config/config.json");
        IngestionAgent agent = new IngestionAgent(config);
        agent.start();
        // Keep the agent running for demonstration (e.g., 1 interval)
        try { Thread.sleep(config.getIngestionIntervalMs() + 2000); } catch (InterruptedException e) {}
    }
}

File: code/agents/ingestion_agent.py
import time
import threading
import requests
from shared.agent_config import AgentConfig
from shared.storage_helper import StorageHelper

class IngestionAgent:
    def __init__(self, config):
        self.config = config
        self.running = False

    def start(self):
        self.running = True
        threading.Thread(target=self._run_loop, daemon=True).start()

    def stop(self):
        self.running = False

    def _run_loop(self):
        while self.running:
            self._ingest_task()
            time.sleep(self.config.get_ingestion_interval_ms() / 1000)

    def _ingest_task(self):
        try:
            print("IngestionAgent: Fetching data from source...")
            url = self.config.get_source_url()
            response = requests.get(url)
            response.raise_for_status()
            
            raw_data = response.text
            StorageHelper.save_to_file(raw_data, self.config.get_storage_output_path())
            print("IngestionAgent: Data ingested and saved.")
        except Exception as e:
            print(f"IngestionAgent: Failed to ingest data: {e}")

if __name__ == "__main__":
    config = AgentConfig("../config/config.json")
    agent = IngestionAgent(config)
    agent.start()
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        agent.stop()
